rollouts/controllers/remoterootsync_controller.go

```diff
@@ -136,42 +141,81 @@ func (r *RemoteRootSyncReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	gkeCluster, err := r.store.GetCluster(ctx, cr.Name)
 	if err != nil {
+		r.updateStatus(ctx, &remoterootsync, clusterConnectivityFailed, nil)
+		return ctrl.Result{}, err
+	}
+
+	readyCondition, found, err := getStatusCondition(gkeCluster.Status.Conditions, "Ready")
+	if !found {
+		r.updateStatus(ctx, &remoterootsync, clusterNotReady, nil)
+		return ctrl.Result{}, err
+	}
+
+	if readyCondition.Status == "False" {
+		r.updateStatus(ctx, &remoterootsync, clusterNotReady, nil)
 		return ctrl.Result{}, err
 	}
 
 	_, dynCl, err := r.store.GetClusterClient(ctx, gkeCluster)
 	if err != nil {
+		r.updateStatus(ctx, &remoterootsync, clusterConnectivityFailed, nil)
 		return ctrl.Result{}, err
 	}
+
 	r.setupWatches(ctx, dynCl, remoterootsync.Name, remoterootsync.Namespace, remoterootsync.Spec.ClusterRef)
 
 	if err := r.patchRootSync(ctx, dynCl, req.Name, &remoterootsync); err != nil {
+		r.updateStatus(ctx, &remoterootsync, clusterConnectivityFailed, nil)
 		return ctrl.Result{}, err
 	}
 
 	syncStatus, err := checkSyncStatus(ctx, dynCl, req.Name)
 	if err != nil {
+		r.updateStatus(ctx, &remoterootsync, clusterConnectivityFailed, nil)
 		return ctrl.Result{}, err
 	}
 
-	if err := r.updateStatus(ctx, &remoterootsync, syncStatus); err != nil {
+	if err := r.updateStatus(ctx, &remoterootsync, "Ready", &syncStatus); err != nil {
 		klog.Errorf("failed to update status: %v", err)
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, syncStatus string) error {
+func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, readyReason string, syncStatus *string) error {
 	logger := log.FromContext(ctx)
 
+	rssReady := metav1.ConditionTrue
+	if readyReason != "Ready" {
+		rssReady = metav1.ConditionFalse
+	}
+
 	// Don't update if there are no changes.
-	if rrs.Status.SyncStatus == syncStatus && rrs.Generation == rrs.Status.ObservedGeneration {
+	syncStatusChanged := syncStatus != nil && rrs.Status.SyncStatus != *syncStatus
+	generationChanged := rrs.Generation == rrs.Status.ObservedGeneration
+	readyStatusChanged := len(rrs.Status.Conditions) == 0 || rrs.Status.Conditions[0].Type != "Ready" || rrs.Status.Conditions[0].Status != rssReady
+
+	if !generationChanged && !syncStatusChanged && !readyStatusChanged {
```

Comment:
Another way to do this will be: compute the new status and compare the new status with current using `reflect.DeepEqual` (that compares the entire struct recursively). That will future proof this code as we evolve the status.

Reply:
Looking into this. I will need to exclude the `Status.Conditions[].LastTransitionTime` field from the comparison as this can vary from the existing status and any status I compute from scratch.